File: challenges/2021/06.js
// Part 1
const solvePuzzle1 = (input) => {
  const fish = parseInput(input);
  return breed(fish, 80);
};

// Part 2
const solvePuzzle2 = (input) => {
  const fish = parseInput(input);
  return breed(fish, 256);
};

const breed = (fish, days) => {
  let state = Array(9).fill(0);

  fish.forEach((f) => state[f - 1]++);

  for (let i = 0; i < days - 1; i++) {
    const newState = state.slice(1).concat(state[0]);
    newState[6] += state[0];
    state = newState;
  }

  return state.reduce((a, b) => a + b, 0);
};

const parseInput = (input) => input.split(",").map((n) => Number(n));

require(__dirname + "/../../utils/test.js").test(__filename, __dirname, solvePuzzle1);
require(__dirname + "/../../utils/test.js").test(__filename, __dirname, solvePuzzle2);

File: challenges/2021/06.exs
Code.compile_file(Path.join([__ENV__.file], ["../../../utils/utils.exs"]))

defmodule Challenge do
  @moduledoc false

  def first_result(input), do: breed(input, 80)

  def second_result(input), do: breed(input, 256)

  defp breed(input, days) do
    state =
      input
      |> Enum.at(0)
      |> String.split(",", trim: true)
      |> Enum.map(&String.to_integer/1)
      |> Enum.reduce(List.duplicate(0, 9), fn val, acc ->
        List.update_at(acc, val - 1, &(&1 + 1))
      end)

    Enum.reduce(1..(days - 1), state, fn _, [q, w, e, r, t, y, u, i, o] ->
      [w, e, r, t, y, u, i + q, o, q]
    end)
    |> Enum.sum()
  end
end

Utils.run(&Challenge.first_result/1, __ENV__.file, "Q1")
Utils.run(&Challenge.second_result/1, __ENV__.file, "Q2")
